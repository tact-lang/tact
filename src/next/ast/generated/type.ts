/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as $c from "@/next/ast/common";
import type * as $ from "@/next/ast/type";

export type Signedness = $.Signedness;
export const allSignedness: readonly $.Signedness[] = ["signed", "unsigned"];
export type IFInt = $.IFInt;
export const IFInt = (
    sign: $.Signedness,
    width: number,
    loc: $c.Range,
): $.IFInt =>
    Object.freeze({
        kind: "FInt",
        sign,
        width,
        loc,
    });
export const isIFInt = ($value: IFInt) => $value.kind === "FInt";
export type VarIntWidth = $.VarIntWidth;
export const allVarIntWidth: readonly $.VarIntWidth[] = ["16", "32"];
export type IFVarInt = $.IFVarInt;
export const IFVarInt = (
    sign: $.Signedness,
    width: $.VarIntWidth,
    loc: $c.Range,
): $.IFVarInt =>
    Object.freeze({
        kind: "FVarInt",
        sign,
        width,
        loc,
    });
export const isIFVarInt = ($value: IFVarInt) => $value.kind === "FVarInt";
export type IntFormat = $.IntFormat;
export type TypeInt = $.TypeInt;
export const TypeInt = (format: $.IntFormat, loc: $c.Range): $.TypeInt =>
    Object.freeze({
        kind: "TyInt",
        format,
        loc,
    });
export const isTypeInt = ($value: TypeInt) => $value.kind === "TyInt";
export type SFBits = $.SFBits;
export const SFBits = (bits: number, loc: $c.Range): $.SFBits =>
    Object.freeze({
        kind: "SFBits",
        bits,
        loc,
    });
export const isSFBits = ($value: SFBits) => $value.kind === "SFBits";
export type SFRemaining = $.SFRemaining;
export const SFRemaining = (loc: $c.Range): $.SFRemaining =>
    Object.freeze({
        kind: "SFRemaining",
        loc,
    });
export const isSFRemaining = ($value: SFRemaining) =>
    $value.kind === "SFRemaining";
export type SFDefault = $.SFDefault;
export const SFDefault = (loc: $c.Range): $.SFDefault =>
    Object.freeze({
        kind: "SFDefault",
        loc,
    });
export const isSFDefault = ($value: SFDefault) => $value.kind === "SFDefault";
export type SliceFormat = $.SliceFormat;
export type TypeSlice = $.TypeSlice;
export const TypeSlice = (format: $.SliceFormat, loc: $c.Range): $.TypeSlice =>
    Object.freeze({
        kind: "TySlice",
        format,
        loc,
    });
export const isTypeSlice = ($value: TypeSlice) => $value.kind === "TySlice";
export type RemFormat = $.RemFormat;
export type TypeCell = $.TypeCell;
export const TypeCell = (format: $.RemFormat, loc: $c.Range): $.TypeCell =>
    Object.freeze({
        kind: "TyCell",
        format,
        loc,
    });
export const isTypeCell = ($value: TypeCell) => $value.kind === "TyCell";
export type TypeBuilder = $.TypeBuilder;
export const TypeBuilder = (
    format: $.RemFormat,
    loc: $c.Range,
): $.TypeBuilder =>
    Object.freeze({
        kind: "TyBuilder",
        format,
        loc,
    });
export const isTypeBuilder = ($value: TypeBuilder) =>
    $value.kind === "TyBuilder";
export type TypeUnit = $.TypeUnit;
export const TypeUnit = (loc: $c.Range): $.TypeUnit =>
    Object.freeze({
        kind: "unit_type",
        loc,
    });
export const isTypeUnit = ($value: TypeUnit) => $value.kind === "unit_type";
export type TypeTensor = $.TypeTensor;
export const TypeTensor = (
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeTensor =>
    Object.freeze({
        kind: "tensor_type",
        typeArgs,
        loc,
    });
export const isTypeTensor = ($value: TypeTensor) =>
    $value.kind === "tensor_type";
export type TypeTuple = $.TypeTuple;
export const TypeTuple = (
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeTuple =>
    Object.freeze({
        kind: "tuple_type",
        typeArgs,
        loc,
    });
export const isTypeTuple = ($value: TypeTuple) => $value.kind === "tuple_type";
export type TypeCons = $.TypeCons;
export const TypeCons = (
    name: $c.TypeId,
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeCons =>
    Object.freeze({
        kind: "cons_type",
        name,
        typeArgs,
        loc,
    });
export const isTypeCons = ($value: TypeCons) => $value.kind === "cons_type";
export type Type = $.Type;
export type TypeMap = $.TypeMap;
export const TypeMap = (key: $.Type, value: $.Type, loc: $c.Range): $.TypeMap =>
    Object.freeze({
        kind: "map_type",
        key,
        value,
        loc,
    });
export const isTypeMap = ($value: TypeMap) => $value.kind === "map_type";
