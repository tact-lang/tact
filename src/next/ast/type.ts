import type { Range, TypeId } from "@/next/ast/common";

export type Type =
    | TypeMap
    | TypeCons
    | TypeInt
    | TypeSlice
    | TypeCell
    | TypeBuilder
    | TypeTuple
    | TypeUnit
    | TypeTensor;

export type TypeCons = {
    readonly kind: "cons_type";
    readonly name: TypeId;
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeInt = {
    readonly kind: "TyInt";
    readonly format: IntFormat;
    readonly loc: Range;
};
export type IntFormat = IFInt | IFVarInt;
export type IFInt = {
    readonly kind: "FInt";
    readonly sign: Signedness;
    readonly width: number;
    readonly loc: Range;
};
export type IFVarInt = {
    readonly kind: "FVarInt";
    readonly sign: Signedness;
    readonly width: VarIntWidth;
    readonly loc: Range;
};
export type VarIntWidth = "16" | "32";
export type Signedness = "signed" | "unsigned";

export type TypeSlice = {
    readonly kind: "TySlice";
    readonly format: SliceFormat;
    readonly loc: Range;
};
export type SliceFormat = SFBits | SFRemaining | SFDefault;
export type SFBits = {
    readonly kind: "SFBits";
    readonly bits: number;
    readonly loc: Range;
};
export type SFRemaining = {
    readonly kind: "SFRemaining";
    readonly loc: Range;
};
export type SFDefault = {
    readonly kind: "SFDefault";
    readonly loc: Range;
};

export type TypeCell = {
    readonly kind: "TyCell";
    readonly format: RemFormat;
    readonly loc: Range;
};
export type TypeBuilder = {
    readonly kind: "TyBuilder";
    readonly format: RemFormat;
    readonly loc: Range;
};
export type RemFormat = SFRemaining | SFDefault;

export type TypeTuple = {
    readonly kind: "tuple_type";
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeUnit = {
    readonly kind: "unit_type";
    readonly loc: Range;
};

export type TypeTensor = {
    readonly kind: "tensor_type";
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeMap = {
    readonly kind: "map_type";
    readonly key: Type; // any type except tensor
    readonly value: Type; // any type except tensor
    readonly loc: Range;
};
