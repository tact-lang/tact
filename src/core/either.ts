// left$0 {X:Type} {Y:Type} value:X = Either X Y;
// right$1 {X:Type} {Y:Type} value:Y = Either X Y;

import { uint } from "@/core/numeric";
import type { Tlb } from "@/core/tlb";

export type Either<L, R> = Left<L> | Right<R>
export type Left<L> = { readonly $: 'left', readonly value: L }
export const Left = <L>(value: L): Left<L> => Object.freeze({ $: 'left', value });
export const isLeft = <L>(m: Either<L, unknown>): m is Left<L> => m.$ === 'left';
export type Right<R> = { readonly $: 'right', readonly value: R }
export const Right = <R>(value: R): Right<R> => Object.freeze({ $: 'right', value });
export const isRight = <R>(m: Either<unknown, R>): m is Right<R> => m.$ === 'right';

export const either = <L, R>(left: Tlb<L>, right: Tlb<R>): Tlb<Either<L, R>> => ({
    store: (t, b, p) => {
        if (t.$ === 'left') {
            uint(1).store(0, b, [...p, '$']);
            left.store(t.value, b, [...p, 'value']);
        } else {
            uint(1).store(1, b, [...p, '$']);
            right.store(t.value, b, [...p, 'value']);
        }
    },
    load: (s, p) => {
        const type = uint(1).load(s, [...p, '$']);
        if (type === 0) {
            return Left(left.load(s, [...p, 'value']));
        } else {
            return Right(right.load(s, [...p, 'value']));
        }
    },
});
