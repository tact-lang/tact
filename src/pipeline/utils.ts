import type { TypeDescription } from "../types/types";

/**
 * This function sorts contracts in topological order.
 * It also checks for cycles in the graph of dependencies and returns undefined if a cycle is found.
 * It works in O(N) time, where N is the number of contracts.
 * @param allContracts - list of all contracts.
 * @returns sorted list of contracts or undefined if a cycle is found
 */
export const topSortContracts = (
    allContracts: TypeDescription[],
): TypeDescription[] | undefined => {
    const visitingNow: Set<TypeDescription> = new Set();
    const visited: Set<TypeDescription> = new Set();
    const result: TypeDescription[] = [];
    const dfs = (contract: TypeDescription): boolean => {
        visitingNow.add(contract);
        for (const c of contract.dependsOn) {
            if (c.kind !== "contract") {
                // This check should also be false. As contracts should be filtered out before calling this function.
                continue;
            }
            if (visitingNow.has(c)) {
                return false;
            }
            if (!visited.has(c)) {
                if (!dfs(c)) {
                    return false;
                }
            }
        }
        visitingNow.delete(contract);
        visited.add(contract);
        result.push(contract);
        return true;
    };
    for (const contract of allContracts) {
        if (!visited.has(contract)) {
            if (!dfs(contract)) {
                return undefined;
            }
        }
    }
    return result;
};
