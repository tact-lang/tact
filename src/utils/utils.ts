import { crc16 } from "./crc16";

export function topologicalSort<T>(src: T[], references: (src: T) => T[]) {
    const result: T[] = [];
    const visited = new Set<T>();
    const visiting = new Set<T>();
    const visit = (src: T) => {
        if (visiting.has(src)) {
            throw Error('Cycle detected');
        }
        if (!visited.has(src)) {
            visiting.add(src);
            for (const r of references(src)) {
                visit(r);
            }
            visiting.delete(src);
            visited.add(src);
            result.push(src);
        }
    }
    for (const s of src) {
        visit(s);
    }
    return result;
}

export function deepFreeze<T>(obj: T) {
    const propNames = Object.getOwnPropertyNames(obj);
    for (const name of propNames) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const value = (obj as any)[name];
        if (value && typeof value === "object") {
            deepFreeze(value);
        }
    }
    return Object.freeze(obj);
}

export function getMethodId(name: string) {
    return (crc16(name) & 0xffff) | 0x10000;
}