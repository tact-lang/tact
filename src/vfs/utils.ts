import { sha256_sync } from "@ton/crypto";
import path from "path";

/**
 * Ensures the resulting file name does not exceed the given maximum length.
 * If too long, trims the name and appends a short hash to avoid collisions.
 *
 * @param name - The base file name without extension.
 * @param ext - The file extension.
 * @param maxLen - Maximum allowed length for the full file name (default: 255).
 * @returns A safe file name within the specified length.
 */
export const makeSafeName = (
    name: string,
    ext: string,
    maxLen = 255,
): string => {
    const full = name + ext;

    if (full.length <= maxLen) {
        return full;
    }

    const hash = sha256_sync(Buffer.from(name)).toString("hex").slice(0, 8);
    const suffix = `_${hash}${ext}`;
    const maxNameLen = maxLen - suffix.length;
    const safeName = name.slice(0, maxNameLen);

    return `${safeName}${suffix}`;
};

/**
 * Returns the full extension of a file, including all parts after the first dot.
 * - "file.txt" => ".txt"
 * - "archive.tar.gz" => ".tar.gz"
 */
export const getFullExtension = (filename: string): string => {
    const base = path.basename(filename);
    const firstDotIndex = base.indexOf(".");
    return firstDotIndex !== -1 ? base.slice(firstDotIndex) : "";
};
