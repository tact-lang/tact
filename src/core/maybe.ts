// nothing$0 {X:Type} = Maybe X;
// just$1 {X:Type} value:X = Maybe X;

import { TlbError, type Path, type Tlb } from "@/core/tlb";
import type { Builder, Slice } from "@ton/core";

export type Maybe<T> = Nothing | Just<T>
export type Nothing = { readonly $: 'nothing' }
export const Nothing: Nothing = Object.freeze({ $: 'nothing' })
export const isNothing = <T>(m: Maybe<T>): m is Nothing => m.$ === 'nothing';
export type Just<T> = { readonly $: 'just', readonly value: T }
export const Just = <T>(value: T): Just<T> => Object.freeze({ $: 'just', value });
export const isJust = <T>(m: Maybe<T>): m is Just<T> => m.$ === 'just';

export const maybe = <T>(type: Tlb<T>): Tlb<Maybe<T>> => {
    const store = (t: Maybe<T>, b: Builder, p: Path) => {
        if (t.$ === 'nothing') {
            b.storeBit(0);
            return;
        }
        try {
            b.storeBit(1);
        } catch (e) {
            throw new TlbError(e, [...p, '$']);
        }
        type.store(t.value, b, [...p, 'value']);
    };
    const load = (s: Slice, p: Path) => {
        const bit = (() => {
            try {
                return s.loadBit();
            } catch (e) {
                throw new TlbError(e, [...p, '$']);
            }
        })();
        if (bit) {
            return Just(type.load(s, [...p, 'value']));
        } else {
            return Nothing;
        }
    };
    return { store, load };
};