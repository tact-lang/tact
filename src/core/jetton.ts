// InternalMsgBody
//     transfer
//         #0f8a7ea5
//         query_id:uint64
//         amount:(VarUInteger 16)
//         destination:MsgAddress
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     transfer_notification
//         #7362d09c
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         forward_payload:(Either Cell ^Cell)
//     excesses
//         #d53276db
//         query_id:uint64
//     burn
//         #595f07bc
//         query_id:uint64
//         amount:(VarUInteger 16)
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//     internal_transfer
//         #178d4519
//         query_id:uint64
//         amount:(VarUInteger 16)
//         from:MsgAddress
//         response_address:MsgAddress
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     burn_notification
//         #7bdd97de
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         response_destination:MsgAddress

import { either } from "@/core/either";
import { maybe } from "@/core/maybe";
import { msgAddress } from "@/core/msg-address";
import { uint, uintVar } from "@/core/numeric";
import { type GetTlb, literal, object, ref, type Tlb } from "@/core/tlb";

export const transfer = <Custom, Forward>(
    custom: Tlb<Custom>,
    forward: Tlb<Forward>,
) => {
    return object({
        $: literal(uint(32))(0x0f8a7ea5),
        queryId: uint(64),
        amount: uintVar(16),
        destination: msgAddress,
        responseDestination: msgAddress,
        customPayload: maybe(ref(custom)),
        forwardTonAmount: uintVar(16),
        forwardPayload: either(forward, ref(forward)),
    });
};

export type Transfer<T, U> = GetTlb<ReturnType<typeof transfer<T, U>>>
