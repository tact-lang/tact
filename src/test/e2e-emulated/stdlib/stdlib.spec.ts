import { Address, beginCell, Cell, toNano } from "@ton/core";
import type { SandboxContract, TreasuryContract } from "@ton/sandbox";
import { Blockchain } from "@ton/sandbox";
import { StdlibTest } from "./output/stdlib_StdlibTest";
import "@ton/test-utils";
import { shouldThrowOnTvmGetMethod } from "@/test/utils/throw";

describe("stdlib", () => {
    let blockchain: Blockchain;
    let treasury: SandboxContract<TreasuryContract>;
    let contract: SandboxContract<StdlibTest>;

    beforeEach(async () => {
        blockchain = await Blockchain.create();
        blockchain.verbosity.print = false;
        treasury = await blockchain.treasury("treasury");

        contract = blockchain.openContract(await StdlibTest.fromInit());

        const deployResult = await contract.send(
            treasury.getSender(),
            { value: toNano("10") },
            null,
        );

        expect(deployResult.transactions).toHaveTransaction({
            from: treasury.address,
            to: contract.address,
            success: true,
            deploy: true,
        });
    });

    it("should execute stdlib methods correctly", async () => {
        const slice = beginCell()
            .storeBit(1)
            .storeBit(1)
            .storeRef(beginCell().storeBit(1).endCell())
            .endCell()
            .asSlice();

        // Execute and verify slice methods
        expect(await contract.getSliceBits(slice)).toBe(2n);
        expect(await contract.getSliceRefs(slice)).toBe(1n);
        expect(await contract.getSliceEmpty(slice)).toBe(false);
        expect(await contract.getLoadBool(slice)).toBe(true);
        expect(await contract.getLoadBit(slice)).toBe(true);

        expect(
            (await contract.getStoreBool(beginCell(), true))
                .endCell()
                .toString(),
        ).toBe(beginCell().storeBit(true).endCell().toString());

        expect(
            (await contract.getStoreBit(beginCell(), true))
                .endCell()
                .toString(),
        ).toBe(beginCell().storeBit(true).endCell().toString());

        expect(Number(await contract.getTvm_2023_07Upgrade())).toMatchSnapshot(
            "tvm_2023_07Upgrade",
        );
        expect(await contract.getTvm_2024_04Upgrade()).toEqual(82009144n);

        expect(
            (
                await contract.getStoreMaybeRef(
                    beginCell(),
                    beginCell().storeUint(123, 64).endCell(),
                )
            ).endCell(),
        ).toEqualCell(
            beginCell()
                .storeMaybeRef(beginCell().storeUint(123, 64).endCell())
                .endCell(),
        );

        expect(
            (await contract.getStoreMaybeRef(beginCell(), null)).endCell(),
        ).toEqualCell(beginCell().storeMaybeRef(null).endCell());

        expect(
            await contract.getLoadMaybeRef(
                beginCell()
                    .storeMaybeRef(beginCell().storeUint(123, 64).endCell())
                    .asSlice(),
            ),
        ).toEqualCell(beginCell().storeUint(123, 64).endCell());

        expect(
            await contract.getLoadMaybeRef(
                beginCell().storeMaybeRef(null).asSlice(),
            ),
        ).toBe(null);

        const addrStd = await contract.getParseStdAddress(
            beginCell()
                .storeAddress(
                    Address.parse(
                        "0:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
                    ),
                )
                .endCell()
                .asSlice(),
        );
        expect(addrStd.workchain).toBe(0n);
        expect(addrStd.address).toBe(
            BigInt(
                "0x4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
            ),
        );

        const forceBasechainGood = await contract.getForceBasechain(
            Address.parse(
                "0:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
            ),
        );
        expect(forceBasechainGood).toBe(true);

        await shouldThrowOnTvmGetMethod(async () => {
            await contract.getForceBasechain(
                Address.parse(
                    "-1:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
                ),
            );
        }, 138);

        const forceWorkchainGoodBasechain = await contract.getForceWorkchain(
            Address.parse(
                "0:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
            ),
            0n,
            123n,
        );
        expect(forceWorkchainGoodBasechain).toBe(true);

        const forceWorkchainGoodMasterchain = await contract.getForceWorkchain(
            Address.parse(
                "-1:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
            ),
            -1n,
            123n,
        );
        expect(forceWorkchainGoodMasterchain).toBe(true);

        await shouldThrowOnTvmGetMethod(async () => {
            await contract.getForceWorkchain(
                Address.parse(
                    "-1:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
                ),
                42n,
                593n,
            );
        }, 593);

        expect(await contract.getBuilderDepth(beginCell())).toBe(0n);
        expect(
            await contract.getBuilderDepth(beginCell().storeRef(Cell.EMPTY)),
        ).toBe(1n);

        expect(await contract.getSkipLastBits(slice, 1n)).toEqualSlice(
            beginCell()
                .storeBit(1)
                .storeRef(beginCell().storeBit(1).endCell())
                .endCell()
                .asSlice(),
        );

        expect(await contract.getFirstBits(slice, 1n)).toEqualSlice(
            beginCell().storeBit(1).endCell().asSlice(),
        );

        expect(await contract.getLastBits(slice, 1n)).toEqualSlice(
            beginCell().storeBit(1).endCell().asSlice(),
        );

        const emptyCell = beginCell().endCell();

        expect(
            await contract.getSkipRef(
                beginCell().storeRef(emptyCell).storeUint(42, 32).asSlice(),
            ),
        ).toEqualSlice(beginCell().storeUint(42, 32).asSlice());

        expect(
            await contract.getSkipMaybeRef(
                beginCell().storeMaybeRef(null).storeUint(42, 32).asSlice(),
            ),
        ).toEqualSlice(beginCell().storeUint(42, 32).asSlice());

        expect(
            await contract.getSkipBool(
                beginCell().storeBit(true).storeUint(42, 32).asSlice(),
            ),
        ).toEqualSlice(beginCell().storeUint(42, 32).asSlice());

        expect(
            await contract.getSkipCoins(
                beginCell().storeCoins(239).storeUint(42, 32).asSlice(),
            ),
        ).toEqualSlice(beginCell().storeUint(42, 32).asSlice());

        expect(
            await contract.getSkipAddress(
                beginCell()
                    .storeAddress(
                        Address.parse(
                            "0:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
                        ),
                    )
                    .storeUint(42, 32)
                    .asSlice(),
            ),
        ).toEqualSlice(beginCell().storeUint(42, 32).asSlice());

        expect(await contract.getSliceDepth(slice)).toBe(1n);

        expect(
            await contract.getComputeDataSizeCell(slice.asCell(), 1000n),
        ).toMatchObject({
            $$type: "DataSize",
            cells: 2n,
            bits: 3n,
            refs: 1n,
        });
        expect(
            await contract.getComputeDataSizeCell(null, 1000n),
        ).toMatchObject({
            $$type: "DataSize",
            cells: 0n,
            bits: 0n,
            refs: 0n,
        });

        expect(
            await contract.getComputeDataSizeSlice(slice, 1000n),
        ).toMatchObject({
            $$type: "DataSize",
            cells: 1n, // -1 for slice
            bits: 3n,
            refs: 1n,
        });

        expect(await contract.getCellDepth(slice.asCell())).toBe(1n);
        expect(await contract.getCellDepth(null)).toBe(0n);

        expect(await contract.getCurLt()).toBe(0n);

        expect(await contract.getBlockLt()).toBe(0n);

        expect(Number(await contract.getSetGasLimit(5000n))).toMatchSnapshot(
            "Gas consumed in segGasLimit()",
        ); // 5000 just to make sure it's enough, 3785 is how much it actually costs
        await expect(contract.getSetGasLimit(3784n)).rejects.toThrow("-14"); // 3784 gas is not enough for sure

        expect(await contract.getGetSeed()).toBe(0n);

        expect(await contract.getSetSeed(123n)).toBe(123n);

        expect(await contract.getMyCode()).toEqualCell(contract.init!.code);

        const varIntegers1 = await contract.getVarIntegers1();
        expect(varIntegers1).toBe(1234n); // 1000 + 200 + 30 + 4

        const varIntegers2 = await contract.getVarIntegers2();
        expect(varIntegers2).toBe(1234n); // 1000 + 200 + 30 + 4
    });

    it("should read forward fee inside receiver correctly", async () => {
        const sendResult = await contract.send(
            treasury.getSender(),
            { value: toNano(1) },
            { $$type: "ReadFwdFeeMsg" },
        );
        expect(sendResult.transactions).toHaveTransaction({
            from: contract.address,
            to: treasury.address,
            success: true,
            body: (body) => {
                if (!body) return false;
                const cs = body.beginParse();
                const fwdFee = cs.loadCoins();
                const expectedFwdFee = cs.loadCoins();
                return fwdFee === expectedFwdFee;
            },
        });
    });
});
