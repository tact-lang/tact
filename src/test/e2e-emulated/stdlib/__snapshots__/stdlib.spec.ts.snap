// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stdlib should execute stdlib methods correctly: Gas consumed in segGasLimit() 1`] = `3827`;

exports[`stdlib should execute stdlib methods correctly: tvm_2023_07Upgrade 1`] = `1385`;
